documentation/client-side-forms/appendix-client-side-form-binding-reference.asciidoc

```diff
@@ -0,0 +1,348 @@
+${
+  binder.for(binder.model.fullName).invalid
+    ? html`
+      <strong>
+        ${binder.for(binder.model.fullName).errors[0]}
+      </strong>`
+    : ''
+}
+----
+
+[NOTE]
+See also: <<appendix-vaadin-components.asciidoc#, Appendix: Binding Data to Vaadin Components>>
+
+== Model
+
+The model is a type-safe reference to a value stored in the binder. Models describe the structure of the data, and are used to read and modify current form values and other related properties and states.
+
+Vaadin automatically generates Model classes for server-side endpoints from Java beans. Usually there is no need to define models manually.
+
+Technically, every model instance represents either a key of a parent model, or the value of the Binder itself (e. g., the form data object).
+
+[NOTE]
+====
+The `Binder` constructor receives the form model class, and takes care of creating model instances. More on that in the `Binder` section below.
+====
+
+//The model constructors receive and store the following arguments:
+//
+//- `parent: ModelParent<T>` — the parent object reference: either the parent model instance, or the `Binder` instance
+//- `key: keyof any` — the key in the parent object to get the value from / set the value to, could be `string` when the parent is an object or `number` when the parent is an array
+//- `...validators: ReadonlyArray<Validator<T>>` — the default validators, typically exported from Java JSR-380 annotations
+
+=== Abstract model
+
+All the models subclass from the `AbstractModel<T>` TypeScript class, where the `T` type argument refers to the value type.
+
+`AbstractModel<T>`, as well as any subclass, define a static method `static createEmptyValue(): T`, that returns the empty value of the subject model type:
+
+[source, typescript]
+----
+const emptyPerson: Person = PersonModel.createEmptyValue();
+console.log(emptyPerson); // {"fullName": ""}
+----
+
+As with any JavaScript object, `AbstractModel` has `toString(): string` and `valueOf(): T` instance methods, that are handy for template expressions:
+
+[soruce, html]
+----
+<p>
+  Full name:
+  ${model.firstName.toString()}
+  ${model.middleName.toString()}
+  ${model.lastName.toString()}
+</p>
+
+<p>
+  Cost:
+   ${model.quantity.valueOf() * model.price.valueOf()}
+</p>
+----
+
+[NOTE]
+====
+The actual data referenced with every model instance, such as the current value, the validators, the required flag, and the validation state, is available on the corresponding `BinderNode` (see below).
+====
+
+=== Primitive models
+
+There are the builtin models that represent the common primitive field types:
+
+|===
+| Type | Value type `T` | Empty value (`Model.createEmptyValue()` result)
+
+| `StringModel` | `string` | `''`
+| `NumberModel` | `number` | `0`
+| `BooleanModel`  | `boolean` | `false`
+|===
+
+Primitive models subclass `PrimitiveModel<T>`. Primitive models are leaf, i. e., do not have nested field keys.
```

Comment:
(minor) I think this would be easier to read and understand. I wouldn't use "subclass" as a verb since it can be confused for a substantive.
```suggestion
Primitive models extend `PrimitiveModel<T>`. Primitive models are leaf nodes of the data structure, that is, they do not have nested field keys.
```